### PR TITLE
tracing: tracer integration with Apache SkyWalking SDK

### LINKED ISSUE #13799: tracing: Apache SkyWalking C++ SDK integration
We have SkyWalking tracer in Envoy. https://github.com/envoyproxy/envoy/pull/13060#pullrequestreview-518395216
I think that this feature should be provided by the external library. So I'm creating C++ SDK for Apache SkyWalking focusing on Envoy integration. https://github.com/SkyAPM/cpp2sky
I'd like to track this movement on this issue. cc @wbpcode @dio @lizan

### PR DESCRIPTION
**Pull request by Shikugawa**
For an explanation of how to fill out the fields, please see the relevant section
in [PULL_REQUESTS.md](https://github.com/envoyproxy/envoy/blob/master/PULL_REQUESTS.md)

Commit Message: Fix #13799. This means integration with external Apache SkyWalking SDK for C++ and destroyed redundant parts of implementations. Hired library is https://github.com/skyAPM/cpp2sky. But it is not released and some of missing parts of implementations which are necessary to achieve clean integration. Currently working private fork of that SDK here. https://github.com/Shikugawa/cpp2sky-envoy. 

Additional Description:

cpp2sky dependency policy check list

### MUST
- [x] Cloud Native Computing Foundation (CNCF) approved license
- [x] Dependencies must not substantially increase the binary size unless they are optional (i.e. confined to specific extensions)
- [x] No duplication of existing dependencies
       Maybe, the dependencies are described [here](https://github.com/SkyAPM/cpp2sky/blob/main/bazel/repositories.bzl)
- [x] Hosted on a git repository and the archive fetch must directly reference this repository. We will NOT support intermediate artifacts built by-hand located on GCS, S3, etc.
- [x] CVE history appears reasonable, no pathological CVE arcs
- [x] Code review (ideally PRs) before merge
- [x] Security vulnerability process exists, with contact details and reporting/disclosure process
- [x] Tests run in CI

### SHOULD
- [ ] > 1 contributor responsible for a non-trivial number of commits
- [ ] High test coverage (also static/dynamic analysis, fuzzing)
      Parts which is used in Envoy have higher test coverage. But gRPC async segment reporter(which is not used in Envoy) has less test.
- [ ] Envoy can obtain advanced notification of vulnerabilities or of security releases
- [ ] Do other significant projects have shared fate by using this dependency?
- [x] Releases (with release notes)
- [x] Commits/releases in last 90 days

Risk Level: Mid
Testing: Unit/Integration
Docs Changes:
Release Notes:
Platform Specific Features:
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Deprecated:]

### CODE REVIEW

**Comment by Shikugawa** [comment]
cc @wbpcode

**Comment by wbpcode** [comment]
That is great work. 👍 
I noticed that the reporter is still using Envoy's own grpc client to report trace data, rather than having the SDK report the trace data automatically.
Is this a final decision or a temporary compromise?

**Comment by wu-sheng** [comment]
> I noticed that the reporter is still using Envoy's own grpc client to report trace data, rather than having the SDK report the trace data automatically.
Is this a final decision or a temporary compromise?

We prefer this. Do you have any concern?

**Comment by Shikugawa** [comment]
@wbpcode @wu-sheng Yes. It is only temporal compromise. I think that cpp2sky's reporter is not matured compared with Envoy's gRPC abstraction. So gradual integration should be better.

File: bazel/repository_locations.bzl
Comment (dio) [inline bazel/repository_locations.bzl:306]: Can we remove this dep?

Comment (Shikugawa) [reply bazel/repository_locations.bzl:306]: cpp2sky depends on data collect protocol so we can't remove this.

File: bazel/repository_locations.bzl
Comment (htuch) [inline bazel/repository_locations.bzl:318]: This is a new dependency, can you please provide an evaluation against the criteria in https://github.com/envoyproxy/envoy/blob/master/DEPENDENCY_POLICY.md#new-external-dependencies? CC @moderation

File: bazel/repository_locations.bzl
Comment (htuch) [reply bazel/repository_locations.bzl:304]: Do any of these APIs appear get used by protos in `api/envoy/`?

File: test/per_file_coverage.sh
Comment (Shikugawa) [inline test/per_file_coverage.sh:55]: skywalking tracer has higier coverage (>= 98.5) but the overall tracer coverage had decreased because remaining 1.5 ratio. So changed this value to 96.3.

**Review by Shikugawa — COMMENTED** [review]
implementation comment

File: bazel/repository_locations.bzl
Comment (Shikugawa) [reply bazel/repository_locations.bzl:304]: No. `skywalking_data_collect_protocol` is completely independent of envoy. Only this tracer dependency depends on these APIs.

File: bazel/repository_locations.bzl
Comment (moderation) [reply bazel/repository_locations.bzl:318]: @htuch Output from [scorecard](https://github.com/ossf/scorecard) is not great (sorted for Pass / Fail). Average score is -0.142857142857143.
```
scorecard --repo=https://github.com/SkyAPM/cpp2sky

RESULTS
-------
Active: Pass 10
CI-Tests: Pass 9
Code-Review: Pass 10
Frozen-Deps: Pass 10
Pull-Requests: Pass 9
Security-Policy: Pass 10

Branch-Protection: Fail 10
CII-Best-Practices: Fail 10
Contributors: Fail 10
Fuzzing: Fail 10
Packaging: Fail 10
SAST: Fail 10
Signed-Releases: Fail 0
Signed-Tags: Fail 0
```

Comment (Shikugawa) [reply bazel/repository_locations.bzl:318]: @moderation Could you show me the average score of scorecard to be valid? I'm a maintainer of this dependency so can do efforts to improve this score.

File: bazel/external/cpp2sky.BUILD
Comment (lizan) [inline bazel/external/cpp2sky.BUILD:1]: This bzl can live in cpp2sky repo instead of here, or you can just bind cpp2sky to @com_github_skyapm_cpp2sky//source:cpp2sky_data_lib

File: examples/verify-common.sh
Comment (lizan) [inline examples/verify-common.sh:42]: ?

File: test/per_file_coverage.sh
Comment (lizan) [inline test/per_file_coverage.sh:57]: You don't need this line.

File: bazel/repository_locations.bzl
Comment (Shikugawa) [reply bazel/repository_locations.bzl:318]: @moderation I cheched `dd-opentracing-cpp` and it results in `-2.5`. So I think it is enough managed.

```
./scorecard --repo=https://github.com/DataDog/dd-opentracing-cpp

RESULTS
-------
CI-Tests: Pass 10
Code-Review: Pass 10
Pull-Requests: Pass 10
Signed-Releases: Pass 10

Active: Fail 10
Branch-Protection: Fail 10
CII-Best-Practices: Fail 10
Contributors: Fail 10
Frozen-Deps: Fail 5
Fuzzing: Fail 10
Packaging: Fail 0
SAST: Fail 10
Security-Policy: Fail 10
Signed-Tags: Fail 0
```

**Comment by Shikugawa** [comment]
@htuch I attached dependency policy requirements on PR description. Could you check it and report about any wondering parts?

File: bazel/repository_locations.bzl
Comment (htuch) [reply bazel/repository_locations.bzl:318]: @Shikugawa you might want to configure branch protection at least, that's easy to do in GH.

File: bazel/repository_locations.bzl
Comment (htuch) [inline bazel/repository_locations.bzl:320]: I had a look at the implementation of this library. It uses its own gRPC async client, built on Google gRPC SDK. I think there are two questions:
1. Does this async client play "nice" with the Envoy thread model? I.e. Do independent worker threads get their own completion queue threads, etc. We have spent some time investing in making this work well inside Envoy's Google gRPC async client.
2. This can only build when we don't have `--define google_grpc=disabled`.

IMHO ideally you'd be able to point your library at our async client and work even when there is no Google gRPC. This is not always possible in 3rd party libraries, but I'd encourage you to explore the practicality.

**Review by htuch — COMMENTED** [review]
Thanks for the updated scorecard. I largely agree except you have `1 contributor responsible for a non-trivial number of commits`, where the requirement is `> 1 contributor responsible for a non-trivial number of commits`.

This seems a reasonable dependency to add; my main concern is around gRPC async clients (see other thread).

File: source/extensions/tracers/skywalking/tracer.h
Comment (lizan) [inline source/extensions/tracers/skywalking/tracer.h:21]: This is a non-POD static initialization and not allowed.

File: bazel/repository_locations.bzl
Comment (lizan) [reply bazel/repository_locations.bzl:320]: I don't think this PR uses the Google gRPC SDK from cpp2sky but only the data model. The Google gRPC based one is more a generic feature for other C++ client using SkyWalking. We might need to be clear in the build target to make sure this doesn't pull google grpc as dependency.

File: bazel/repository_locations.bzl
Comment (Shikugawa) [reply bazel/repository_locations.bzl:318]: @htuch Branch protection has already hired on cpp2sky. It is only protection to push main branch directly. Is it not enough?

File: bazel/repository_locations.bzl
Comment (Shikugawa) [reply bazel/repository_locations.bzl:320]: Yes. It uses only Skywalking's data definition. gRPC client abstraction has already accomplished by integration envoy internal grpc client. For more detail, https://github.com/envoyproxy/envoy/blob/a93219b052b6cd69f99dcb66f5c99c99b2cc072b/source/extensions/tracers/skywalking/trace_segment_reporter.cc

Comment (htuch) [reply bazel/repository_locations.bzl:320]: I see, yep, makes sense to me then. I think as long as this is passing `compile_time_options` it probably works without gRPC. Adding some comments around this would be helpful.

**Comment by htuch** [comment]
Re: branch protection; does OSSF Scorecard now pass its check?

**Comment by Shikugawa** [comment]
@htuch cpp2sky's scorecard result,

```
RESULTS
-------
Active: Pass 10
CI-Tests: Pass 9
Code-Review: Pass 10
Frozen-Deps: Pass 10
Security-Policy: Pass 10

Branch-Protection: Fail 4
CII-Best-Practices: Fail 10
Contributors: Fail 10
Fuzzing: Fail 10
Packaging: Fail 10
SAST: Fail 10
Signed-Releases: Fail 0
Signed-Tags: Fail 0
```

Branch protection score seems Fail. But, envoy one seems also failed,

```
RESULTS
-------
Active: Pass 10
CI-Tests: Pass 9
CII-Best-Practices: Pass 10
Code-Review: Pass 10
Contributors: Pass 10
Pull-Requests: Pass 10
SAST: Pass 10
Fuzzing: Pass 10
Security-Policy: Pass 10

Branch-Protection: Fail 10
Packaging: Fail 10
Frozen-Deps: Fail 5
Signed-Releases: Fail 0
Signed-Tags: Fail 0
```

**Comment by htuch** [comment]
@dlorenc @inferno-chromium any idea why Envoy branch protection is failing on scorecards for branch protection? I'm pretty sure we have a fairly conservative setup here.

**Comment by inferno-chromium** [comment]
> @dlorenc @inferno-chromium any idea why Envoy branch protection is failing on scorecards for branch protection? I'm pretty sure we have a fairly conservative setup here.

Can you check https://github.com/envoyproxy/envoy/settings/branches and verify that you enabled it on the main branch. from scorecard, running with --show-details, it looks like error is https://api.github.com/repos/envoyproxy/envoy/branches/main/protection. Branch protection is not enabled by default and in a kind of hidden place (not in any security tab), so most people forget / unaware about this.

image

**Comment by lizan** [comment]
@inferno-chromium  The branch protection has been there for quite a while, I double checked and it is protected for sure.

**Comment by inferno-chromium** [comment]
> @inferno-chromium The branch protection has been there for quite a while, I double checked and it is protected for sure.

Maybe this branch protection check requires from be run from repo admin. @lizan are you one, can you try running scorecard yourself to see if it works.

**Comment by naveensrinivasan** [comment]
> > @inferno-chromium The branch protection has been there for quite a while, I double checked and it is protected for sure.
> 
> Maybe this branch protection check requires from be run from repo admin. @lizan are you one, can you try running scorecard yourself to see if it works. @naveensrinivasan fyi

Created a tracking issue for this https://github.com/ossf/scorecard/issues/138

**Comment by Shikugawa** [comment]
@htuch Now it seems quite enough cpp2sky's branch protection score following with 
 https://github.com/envoyproxy/envoy/pull/14493#issuecomment-766533926

```
RESULTS
-------
Active: Pass 10
Branch-Protection: Pass 10
CI-Tests: Pass 9
Code-Review: Pass 10
Frozen-Deps: Pass 10
Pull-Requests: Pass 9
Security-Policy: Pass 10

CII-Best-Practices: Fail 10
Contributors: Fail 10
Fuzzing: Fail 10
Packaging: Fail 10
SAST: Fail 10
Signed-Releases: Fail 0
Signed-Tags: Fail 0
```

File: bazel/repository_locations.bzl
Comment (htuch) [reply bazel/repository_locations.bzl:320]: If you can add a comment here we can close this thread out. Other than this, I'm good to go based on your dependency analysis. @envoyproxy/dependency-shepherds any additional asks here?

**Review by htuch — COMMENTED** [review]
I took the liberty of fixing the " > 1 contributor responsible for a non-trivial number of commits" in the commit message. Otherwise LGTM.

**Comment by Shikugawa** [comment]
@lizan ping

**Review by lizan — APPROVED** [review]

**Pull request merged**
